raft.go

```diff
@@ -1232,23 +1238,22 @@ func (r *Raft) quorumSize() int {
 	return voters/2 + 1
 }
 
-// appendConfigurationEntry changes the configuration and adds a new
-// configuration entry to the log
-func (r *Raft) appendConfigurationEntry(future *configurationChangeFuture) {
-	configuration := cloneConfiguration(r.configurations.latest) // same as committed
-
-	if future.prevIndex > 0 && future.prevIndex != r.configurations.latestIndex {
+// nextConfiguration generates a new Configuration from the current one and a
+// configuration change request. It's split from appendConfigurationEntry so
+// that it can be unit tested easily. This function reports errors directly to
+// the future; it returns an empty configuration if and only if
+// future.responded is true.
+func nextConfiguration(current Configuration, currentIndex uint64, future *configurationChangeFuture) Configuration {
```

Comment:
I guess. The signature is uglier than I hoped. Maybe the future should have a nested struct called `req`, and this should take a `req` and return a `(Configuration, error)`?

The other ugly thing is that we can't log in this function. But overall it's better than pulling in all the test setup of `raft_test.go`.